mgr/dashboard: test prometheus rules through promtool

aaSharma14 opened the pull request:
This PR intends to add unit testing for prometheus/alertmanager rules using promtool. To run the tests- With ceph_default_alerts.yml in the same directory, run 'promtool test rules test.yml'

Fixes: https://tracker.ceph.com/issues/45415
Signed-off-by: Aashish Sharma <aasharma@redhat.com>



<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

alfonsomthd commented:
@aaSharma14 I think that in order to complete this work it is needed a script like (just an example):

https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/run-frontend-e2e-tests.sh

so that test you mention in the description can be run (and even create a jenkins job in case we find it necessary).

aaSharma14 commented:
> @aaSharma14 I think that in order to complete this work it is needed a script like (just an example):
> 
> https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/run-frontend-e2e-tests.sh
> 
> so that test you mention in the description can be run (and even create a jenkins job in case we find it necessary).

Thanks @alfonsomthd ..will work on the script.

p-se commented on monitoring/prometheus/alerts/ceph_default_alerts.yml:
```diff
@@ -215,7 +215,7 @@ groups:
 
       - alert: MTU Mismatch
-        expr: node_network_mtu_bytes{device!="lo"} != on(device) group_left() (quantile(0.5, node_network_mtu_bytes{device!="lo"}) by (device))
+        expr: node_network_mtu_bytes{device!="lo"} != on() group_left() (quantile(0.5, node_network_mtu_bytes{device!="lo"}))
```
Is this change supposed to be here?

aaSharma14 replied:
Yes, while writing the unit test for this alert, I found out that this change should be done in order to get the correct alert. I have created a separate PR for this change (https://github.com/ceph/ceph/pull/39462). But in order to run these tests successfully, I did this change in this PR as well.

p-se reviewed (commented):
This is a great addition. It seems to work very good. Thanks for that!

If you could rename the commit message to not include "[WIP]" anymore, I'd be glad to approve.

Additionally, but also optionally, it'd be good to have consistent indentation in the `test_alerts.yml` file.

p-se commented on src/pybind/mgr/dashboard/run-promtool-unittests.sh:
```diff
@@ -0,0 +1,13 @@
+    -v "$CEPH_ROOT":/ceph \
+    --name=promtool \
+    --network=host \
+    dnanexus/promtool:2.9.2 \
```
I haven't checked, but maybe there's a way to get promtool from the Prometheus folks directly? I'd prefer that.

epuertat replied:
@p-se unfortunately Prometheus is not widely packaged as it seems that containerized deployment is the preferred way...

In any case, I'd move this to the `install-deps.sh` stage, together with packages, pip, etc.

BTW this might be troublesome due to Dockerhub pull limits. @djgalloway @tchaikov what do you think about this? Any suggestion?

p-se replied:
@epuertat My bad. What I meant was this:

```
docker run --rm -it --entrypoint promtool prom/prometheus:v2.9.2
```

djgalloway replied:
All of the Jenkins builders inside the Ceph lab are configured to try and internal dockerhub mirror first so it might be okay.

aaSharma14 commented:
> This is a great addition. It seems to work very good. Thanks for that!
> 
> If you could rename the commit message to not include "[WIP]" anymore, I'd be glad to approve.
> 
> Additionally, but also optionally, it'd be good to have consistent indentation in the `test_alerts.yml` file.

Thanks @p-se . I have added consistent indentation in the test_alerts.yml file and changed the commit message as well.

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

epuertat commented on src/pybind/mgr/dashboard/run-promtool-unittests.sh:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+set -e
+
+SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+: ${CEPH_ROOT:=$SCRIPTPATH/../../../../}
+
+docker run --rm \
+    -v "$CEPH_ROOT":/ceph \
+    --name=promtool \
+    --network=host \
+    dnanexus/promtool:2.9.2 \
+    test rules /ceph/monitoring/prometheus/alerts/test_alerts.yml
```
I'd rather put this into the dashboard tox file. We were already using tox for checking Grafana dashboards (`check` toxenv).

epuertat replied:
Minor: instead of mounting a volume you can simply pipe the file to the container (-i required):
```suggestion
cat $CEPH_ROOT/monitoring/prometheus/alerts/test_alerts.yml | \
    docker run --rm -i \
    dnanexus/promtool:2.9.2 \
    test rules /dev/stdin
```

epuertat reviewed (changes requested):
My main suggestion is to explicitly pull that docker image during the `install-deps.sh` stage, so we allow the `make check` process to run without internet connection.

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
@alfonsomthd @epuertat I have added a new test to the make check `run-promtool-unittetsts.sh`. I tried pulling the promtool container image in the `intsall-deps.sh`, the image gets pulled but still the test fails with `promtool:command not found`.

p-se commented on install-deps.sh:
```diff
@@ -13,6 +13,7 @@
 #  version 2.1 of the License, or (at your option) any later version.
 #
 set -e
+docker run --rm -it --entrypoint promtool prom/prometheus:v2.9.2
```
This should presumably be `docker pull prom/prometheus:v2.9.2`. `docker run --rm -it ...` does also pull the image but also runs the container with the `promtool` executable and no arguments to `promtool`. The purpose would be clearer if it'd just pull, IMHO.

aaSharma14 replied:
Will do that. Thanks

p-se commented on src/test/run-promtool-unittests.sh:
```diff
@@ -0,0 +1,5 @@
+
+set -e
+
+promtool test rules /ceph/monitoring/prometheus/alerts/test_alerts.yml
```
At this point it should probably then try to run the test using `promtool` of the container like so:

```bash
docker run --rm -it --entrypoint promtool prom/prometheus:v2.9.2 test rules PATH
```

where PATH is the path to the `test_alerts.yml` file inside the container. For this to succeed, `promtool` would also need to find `ceph_default_alerts.yml` next to `test_alerts.yml` inside the container. If it isn't already in the container (which, according to the state of this pull request,  I do not _assume_ is the case), then you'd need to map the file to the container when starting the container with `-v SOURCE_FILE:TARGET_FILE:ro`. Although promtool should not change the content of any file mapped to the container, it might still be appropriate to set the file to read-only using the `:ro` suffix.

So, you'll probably end up with something like the following. Note that I have not tested this code.

```bash
docker run --rm -it \
  --entrypoint promtool \
  -v /ceph/monitoring/prometheus/alerts:/alerts:ro \
  test rules /alerts/test_alerts.yml
```

aaSharma14 replied:
Thanks @p-se ..But the docker command is also not getting executed inside the script..It shows `docker:command not found`.

aaSharma14 replied:
By pulling the pormtool container in the `install-deps.sh`, I could see the traces of the container in the logs..but still got `promtool: command not found` error in the make-check.

p-se reviewed (commented):
> @alfonsomthd @epuertat I have added a new test to the make check `run-promtool-unittetsts.sh`. I tried pulling the promtool container image in the `intsall-deps.sh`, the image gets pulled but still the test fails with `promtool:command not found`.

My comments on `src/test/run-promtool-unittests.sh ` might help resolving this issue.

epuertat reviewed (approved):
Looks good now @aaSharma14 !

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

avanthakkar commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

tchaikov commented on src/test/run-promtool-unittests.sh:
```diff
@@ -0,0 +1,13 @@
+         --name=promtool \
+         --network=host \
+         dnanexus/promtool:2.9.2 \
+         test rules /ceph/monitoring/prometheus/alerts/test_alerts.yml
```
i think this should not be a part of "make check" which is supposed to be composed of unit test and a very basic smoke test.

tchaikov replied:
could you put it in `qa/suites/rados/mgr` ?

epuertat replied:
@tchaikov this is actually unit testing (it feeds the Alerting rules with fake data and test their outcomes) and does not require any Ceph code compiled and/or running. The only 'complexity' here is how to run the `promtool` binary: @aaSharma14 already tried with distro packages, but it seems that the version of `promtool` packaged in Ubuntu 16.04 doesn't support the 'test rules' option :/

The only alternative I can see here is using a Github action to directly run this container image... but I'm not too fond of adding GH actions for unit-testing...

badone commented:
I struck the same issue as rishabh-d-dave.

Why are we installing docker on systems where podman already exists? The reason I run podman is so I don't have to install/run docker and all that entails.

aaSharma14 commented:
> I struck the same issue as rishabh-d-dave.
> 
> Why are we installing docker on systems where podman already exists? The reason I run podman is so I don't have to install/run docker and all that entails.

Hey @badone . This is being tracked in this PR - https://github.com/ceph/ceph/pull/39995
